File: OpenSim Tools/ProcessingTrialDataFromVicon/remove_bad_markers.py
import numpy as np

def remove_bad_markers(trimmed_markers: dict, marker_names: list):
	'''
	This function checks for gaps and duplicate markers, then removes them.

	'''

	# Check for gaps
	bad_keys = []
	
	for key in trimmed_markers.keys():
		if np.size(np.where(abs(trimmed_markers[key]["X"]) == 0)) > 10:
			bad_keys.append(key)
	
	# Cannot get duplicate keys in a dictionary structure

	# Remove all bad markers, if there is a duplicate, remove both
	for key in bad_keys:
		trimmed_markers.pop(key, None)
		marker_names.remove(key)

	return trimmed_markers, marker_names, bad_keys

File: ProcessingTrialDataFromVicon_MATLAB/removeBadMarkers.m
function [goodMarkers,goodMarkerNames, badMarkerNames] = removeBadMarkers(trimmedMarkers,markerNames)
%check for gaps
badInds = [];
for i = 1:length(trimmedMarkers)
   if length(find(abs(trimmedMarkers(i).x) == 0)) > 10
      badInds =  [badInds i];
   end
end

%check for duplicates
[~,uniqueID] = unique(markerNames);
duplicateMarkers = markerNames;
duplicateMarkers(uniqueID) = [];

if ~isempty(duplicateMarkers)
    for i = 1:length(duplicateMarkers)
        dupeMarkerInds = find(contains(markerNames,duplicateMarkers{i}));
        badInds = [badInds dupeMarkerInds(2:end)];
    end
end

%remove all bad markers
onesvector = ones(length(trimmedMarkers),1);
onesvector(badInds) = 0;
goodMarkers = trimmedMarkers(find(onesvector==1));
goodMarkerNames = markerNames(find(onesvector==1));
badMarkerNames = markerNames(find(onesvector~=1));    
end